docs/listeners/jsonapi.rst

```diff
@@ -0,0 +1,723 @@
+    foreach (API_RESOURCES as $apiResource) {
+        $routes->resources($apiResource, [
+            'inflect' => 'dasherize'
+        ]);
+	}
+  }
+
+Controller
+^^^^^^^^^^
+
+Attach the listener using the components array if you want to attach
+it to all controllers, application wide and make sure ``RequestHandler``
+is loaded **before** ``Crud``.
+
+.. code-block:: php
+
+  <?php
+  class AppController extends Controller {
+
+    public function initialize()
+    {
+      $this->loadComponent('RequestHandler');
+      $this->loadComponent('Crud.Crud', [
+        'actions' => [
+          'Crud.Index',
+          'Crud.View'
+        ],
+        'listeners' => ['Crud.JsonApi']
+      ]);
+    }
+
+
+Alternatively, attach the listener to your controllers ``beforeFilter`` 
+if you prefer attaching the listener to specific controllers on the fly.
+
+.. code-block:: php
+
+  <?php
+  class SamplesController extends AppController {
+
+    public function beforeFilter(\Cake\Event\Event $event) {
+      parent::beforeFilter();
+      $this->Crud->addListener('Crud.JsonApi');
+    }
+  }
+
+
+CakeRequest detector
+--------------------
+
+The JsonApi Listener adds the ``is(jsonapi)`` request detector 
+to your ``CakeRequest`` object which check if the request
+contains a ``HTTP Accept`` header  set to ``application/vnd.api+json``.
+
+.. note::
+
+To make sure the listener won't get in your way it will 
+return ``NULL`` for all requests unless ``is('jsonapi')`` is true.
+
+Exception handler
+-----------------
+
+The JsonApi listener overrides the ``Exception.renderer`` for ``jsonapi`` requests,
+so in case of an error, a standardized error in JSON API format will be returned
+for both errors/exceptions and validation errors.
+
+Errors/exceptions
+^^^^^^^^^^^^^^^^^
+
+For standard errors (e.g. 404) and exceptions the listener will
+produce error responses in the following JSON API format:
+
+.. code-block:: json
+
+  {
+    "errors": [
+      {
+        "code": "501",
+        "title": "Not Implemented"
+      }
+    ],
+    "debug": {
+      "class": "Cake\\Network\\Exception\\NotImplementedException",
+      "trace": []
+    }
+  }
+
+.. note::  
+
+Please note that the ``debug`` node with the stack trace will only be included if ``debug`` is true.
+  
+
+Validation errors
+^^^^^^^^^^^^^^^^^
+
+For (422) validation errors the listener produces will produce
+validation error reponses in the following JSON API format.
+
+.. code-block:: json
+
+  {
+    "errors": [
+	  {
+	    "title": "_required",
+	    "detail": "Primary data does not contain member 'type'",
+	    "source": {
+		  "pointer": "/data"
+	    }
+	  }
+    ]
+  }
+ 
+.. note::  
+
+Please note that the listener also responds with (422) validation errors
+when data is posted in a document structure that does not comply with the
+JSON API specification.
+
+
+Response formats
+----------------
+
+HTTP GET (index)
+^^^^^^^^^^^^^^^^
+
+Requests to the ``index`` action **must** use:
```

Comment:
What would the url look like here? `example.com/countries` or `example.com/countries.json` or with `index.json` ?